agents/mcts.c

```diff
@@ -0,0 +1,158 @@
+#include <assert.h>
+#include <float.h>
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "game.h"
+#include "mcts.h"
+#include "util.h"
+
+struct node {
+    int move;
+    char player;
+    int n_visits;
+    double score;
+    struct node *parent;
+    struct node *children[N_GRIDS];
+};
+
+static struct node *new_node(int move, char player, struct node *parent)
+{
+    struct node *node = malloc(sizeof(struct node));
+    node->move = move;
+    node->player = player;
+    node->n_visits = 0;
+    node->score = 0;
+    node->parent = parent;
+    memset(node->children, 0, sizeof(node->children));
+    return node;
+}
+
+static void free_node(struct node *node)
+{
+    for (int i = 0; i < N_GRIDS; i++)
+        if (node->children[i])
+            free_node(node->children[i]);
+    free(node);
+}
+
+static inline double uct_score(int n_total, int n_visits, double score)
+{
+    if (n_visits == 0)
+        return DBL_MAX;
+    return score / n_visits + C * sqrt(log(n_total) / n_visits);
+}
+
+static struct node *select_move(struct node *node)
+{
+    struct node *best_node = NULL;
+    double best_score = -1;
+    for (int i = 0; i < N_GRIDS; i++) {
+        if (node->children[i]) {
+            double score =
+                uct_score(node->n_visits, node->children[i]->n_visits,
+                          node->children[i]->score);
+            if (score > best_score) {
+                best_score = score;
+                best_node = node->children[i];
+            }
+        }
+    }
+    return best_node;
+}
+
+static double simulate(char *table, char player)
+{
+    char win;
+    char current_player = player;
+    char temp_table[N_GRIDS];
+    memcpy(temp_table, table, N_GRIDS);
+    while (1) {
+        int *moves = available_moves(temp_table);
+        if (moves[0] == -1) {
+            free(moves);
+            break;
+        }
+        int n_moves = 0;
+        while (n_moves < N_GRIDS && moves[n_moves] != -1)
+            ++n_moves;
+        int move = moves[rand() % n_moves];
+        free(moves);
+        temp_table[move] = current_player;
+        if ((win = check_win(temp_table)) != ' ') {
+            if (win == player)
+                return 1;
+            if (win == (player ^ 'O' ^ 'X'))
```

Comment:
I like this trick to decide the winner. Can you provide the dedicated helper function?